mysql/mysqldump.go

```diff
@@ -0,0 +1,578 @@
+	isPk := false
+	for _, elem := range col.Options {
+		switch op := elem.Tp; op {
+		case ast.ColumnOptionPrimaryKey:
+			column.NotNull = true
+			// If primary key is defined in a column then `isPk` will be true
+			// and this column will be added in colDef as primary keys
+			isPk = true
+		case ast.ColumnOptionNotNull:
+			column.NotNull = true
+		case ast.ColumnOptionDefaultValue:
+			column.Ignored.Default = true
+		case ast.ColumnOptionUniqKey:
+			column.Unique = true
+		case ast.ColumnOptionCheck:
+			column.Ignored.Check = true
+		case ast.ColumnOptionReference:
+			column.Ignored.ForeignKey = true
+		default:
+			conv.Unexpected(fmt.Sprintf("Not supported type option of const value `%v` found in table : %s , column : %s. Mapping of columnOptionType can be found here: 'https://godoc.org/github.com/pingcap/parser/ast#ColumnOptionType'", op, tableName, column.Name))
+		}
+	}
+	return isPk
+}
+
+// getTypeModsAndID returns ID and mods of column datatype
+func getTypeModsAndID(columnType string) (string, []int64) {
+	// There are no methods in pincap parser to retirieve ID and mods
+	// We will process columnType eg:'varchar(40)' and split ID from the string.
+	// We retrieve mods using regex expression and convert it to INT64
+	id := columnType
+	var mods []int64
+	if strings.Contains(columnType, "(") {
+		id = strings.Split(columnType, "(")[0]
+		// For 'set' and 'enum' datatypes, values provided are not maxLength
+		if id == "set" || id == "enum" {
+			return id, nil
+		}
+		values := valuesRegexp.FindString(columnType)
+		strMods := strings.Split(values[1:len(values)-1], ",")
+		for _, i := range strMods {
+			j, err := strconv.ParseInt(i, 10, 64)
+			if err != nil {
+				panic(err)
+			}
+			mods = append(mods, j)
+		}
+	}
+	// 'BINARY' keyword will be added to all blob datatypes by parser.
+	// Eg: mediumblob BINARY. It needs to be ignored to retrieve ID
+	if strings.Contains(id, " ") {
+		id = strings.Split(columnType, " ")[0]
+	}
+	return id, mods
+}
+
+// handleParseError handles error while parsing mysqldump
+// statements and attempts at creating parsable query.
+// Error can be due to insert statement or unsupported Spatial
+// datatypes in create statement.
+func handleParseError(conv *internal.Conv, query string, err error, l [][]byte) ([]ast.StmtNode, bool) {
+	// Check if error is due to Insert statement.
+	insertStmt := insertRegexp.FindString(query)
+	if insertStmt != "" {
+		// Likely causes of failing to parse Insert statement:
+		// a) Due to some invalid value
+		// b) Query size is more than what pingcap parser could handle (more than 40MB in size)
+		return handleInsertStatement(conv, query, insertStmt)
+	}
+
+	// Check if error is due to spatial datatype as it is not supported by Pingcap parser.
+	var isSpatial bool
+	for _, spatial := range MysqlSpatialDataTypes {
+		if strings.Contains(strings.ToLower(err.Error()), spatial) && isSpatial == false {
```

Comment:
Suggest dropping the "&& isSpatial == false"
and just putting a break at the end of the if block.
Once you've found one spatial type in err, we can stop looking.

Reply:
Sure, I have dropped `isSpatial == false` and replaced it with break.

> Actually, it would be even clearer and simpler if you just called handleSpatialDatatype(conv, query, l)
> instead of the break.
> That way you don't need the isSpatial variable at all, and you don't need the extra conditional after the for loop.

If we drop isSpatial, then we would just be parsing the chunk for `create table..` statement again which is fine if there is a spatial datatype used in the chunk. If there is actually no spatial datatype and we have also dropped isSpatial bool, then it might be due to some different error and we need to `return false` from `handleParseError` and handle such case by reading next line and trying again and adding it as reparsed statement in `readAndParseChunk`. But, what will happen is that `handleSpatialDatatype(conv, query, l)` will again parse the same chunk and if still there is an error then it would be skipping that chunk which is wrong.

I am not sure if I am able to convey this properly but I hope you understood this point. Let me know.

Reply:
Not sure I do follow.

Why isn't:

	for _, spatial := range MysqlSpatialDataTypes {
		if strings.Contains(strings.ToLower(err.Error()), spatial) {
			isSpatial = true
			if conv.SchemaMode() {
				conv.Unexpected(fmt.Sprintf("....", spatial, len(l), chunk))
				internal.VerbosePrintf("Converting datatype ....", spatial)
			}
			break
		}
	}
	if isSpatial {
		return handleSpatialDatatype(conv, strings.ToLower(chunk), l)
	}

identical to:

	for _, spatial := range MysqlSpatialDataTypes {
		if strings.Contains(strings.ToLower(err.Error()), spatial) {
			if conv.SchemaMode() {
				conv.Unexpected(fmt.Sprintf("....", spatial, len(l), chunk))
				internal.VerbosePrintf("Converting datatype ....", spatial)
			}
			return handleSpatialDatatype(conv, strings.ToLower(chunk), l)
		}
	}

In other words:
a) looking for spatial, setting a boolean and breaking from the loop, and then immediately using the boolean to determine whether to return handleSpatialDatatype(...), and
b) just replacing the break with the return handleSpatialDatatype(...)
should be the same?

Reply:
There could be 2 scenarios for error while parsing create table statement:
a) Error could be due to pingcap parser - unable to parse spatial datatype
b) Error could be due to complex statements with embedded semicolons for which we already have a logic in place where we do `conv.Stats.Reparsed++` and deal with this case by reading another line and trying again.

- Consider that error is due to spatial datatype, then `handleSpatialDatatype()` will try to replace all instances of spatial datatype with `TEXT` and retry parsing the statement through pingcap parser. If there is still an error then we are skipping the statement `conv.SkipStatement("Unparsable statement")` and return `nil, true` which means that chunks will not be reparsed.

- On the other hand, if there is no spatial datatype present in the chunk [isSpatial=false], then we can simply `return nil, false` which means our code will try to read next lines and reparse. This case would be ignored if we follow the 2nd approach or we would need to add isSpatial inside the `handleSpatialDatatype()` function and return `nil, false`.

**But if we assume that by replacing spatial datatypes with `TEXT` will solve errors on handling spatial datatype which is a pretty fair assumption, then we can directly `return nil, false` which then would be the case for reparsing and reading next lines and remove the `conv.SkipStatement("unparsable statement")` part.**

**Conclusion :  I will go ahead and remove isSpatial bool and make changes as suggested by you along with some other changes.**

Final code:
```
for _, spatial := range MysqlSpatialDataTypes {
		if strings.Contains(strings.ToLower(err.Error()), spatial) {
			isSpatial = true
			if conv.SchemaMode() {
				conv.Unexpected(fmt.Sprintf("Unsupported datatype '%s' encountered while parsing following statement at line number %d : \n%s", spatial, len(l), chunk))
				internal.VerbosePrintf("Converting datatype '%s' to 'Text' and retrying to parse the statement\n", spatial)
			}
			return handleSpatialDatatype(conv, strings.ToLower(chunk), l)
		}
	}
```
```
func handleSpatialDatatype(conv *internal.Conv, chunk string, l [][]byte) ([]ast.StmtNode, bool) {
	if !conv.SchemaMode() {
		return nil, true
	}
	for _, spatial := range MysqlSpatialDataTypes {
		chunk = strings.ReplaceAll(chunk, " "+spatial, " text")
	}
	newTree, _, err := parser.New().Parse(chunk, "", "")
	if err != nil {
		[REMOVED]//conv.SkipStatement("Unparsable statement")
		[UPDATED TO FALSE]return nil, false
	}
	return newTree, true
}
```

Reply:
Apologies for the long explanation. Just wanted to let you know all the scenarios. Although, this would be like a super rare condition of having both together: 

- usage of spatial datatype.

- complex statement with embedded semicolon in create table statement

Reply:
Done. Resolving this conversation.